hail/src/main/scala/is/hail/io/tabix/TabixReader.scala

```diff
@@ -0,0 +1,355 @@
+package is.hail.io.tabix
+
+import is.hail.HailContext
+import is.hail.io.compress.BGzipInputStream
+import is.hail.utils._
+
+import htsjdk.tribble.util.{ParsingUtils, TabixUtils}
+import org.apache.{hadoop => hd}
+import org.apache.hadoop.io.compress.SplitCompressionInputStream
+import org.apache.spark.SparkContext
+
+import java.io.InputStream
+import java.nio.{ByteBuffer, ByteOrder}
+import java.util.Arrays
+
+import scala.collection.mutable.HashMap
+import scala.language.implicitConversions
+
+// Helper data classes
+
+case class Tabix(
+  val format: Int,
+  val colSeq: Int,
+  val colBeg: Int,
+  val meta: Int,
+  val seqs: Array[String],
+  val chr2tid: HashMap[String, Int],
+  val indices: Array[(HashMap[Int, Array[TbiPair]], Array[Long])]
+)
+
+case class TbiPair(var _1: Long, var _2: Long) extends java.lang.Comparable[TbiPair] {
+  @Override
+  def compareTo(other: TbiPair) = TbiOrd.compare(this, other)
+}
+
+object TbiPair {
+  implicit def tup2Pair(t: (Long, Long)): TbiPair = new TbiPair(t._1, t._2)
+}
+
+object TbiOrd extends Ordering[TbiPair] {
+  def compare(u: TbiPair, v: TbiPair) = if (u._1 == v._1) {
+    0
+  } else if (less64(u._1, v._1)) {
+    -1
+  } else {
+    1
+  }
+
+  def less64(u: Long, v: Long) = (u < v) ^ (u < 0) ^ (v < 0)
+}
+
+// Tabix reader
+// Tabix file format is described here: https://samtools.github.io/hts-specs/tabix.pdf
+
+object TabixReader {
+  val MaxBin: Int = 37450
+  val TadLidxShift: Int = 14
+  val DefaultBufferSize: Int = 1000
+  val Magic: Array[Byte] = Array(84, 66, 73, 1) // "TBI\1"
+
+  def readInt(is: InputStream): Int =
+    (is.read() & 0xff) |
+    ((is.read() & 0xff) << 8) |
+    ((is.read() & 0xff) << 16) |
+    ((is.read() & 0xff) << 24)
+
+  def readLong(is: InputStream): Long =
+    (is.read() & 0xff).asInstanceOf[Long] |
+    ((is.read() & 0xff).asInstanceOf[Long] << 8) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 16) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 24) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 32) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 40) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 48) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 56)
+
+  def readLine(is: InputStream): String = readLine(is, DefaultBufferSize)
+
+  def readLine(is: InputStream, initialBufferSize: Int): String = {
+    val buf = new StringBuffer(initialBufferSize)
+    var c = is.read()
+    while (c >= 0 && c != '\n') {
+      buf.append(c.asInstanceOf[Char])
+      c = is.read()
```

Comment:
`read` returns one byte, so this definitely won't read UTF-8 or UTF-16 correctly. The linked tabix PDF does not document what `char` means or the encoding of `names`.

Reply:
thank you @hail-ci-test-1

Reply:
😁 my bad

Reply:
So `readLine` isn't reading a tabix file, it is reading the indexed file, e.g., VCF.  Earlier versions of VCF also made no comment on encoding, but VCF 4.3 specifies UTF-8 with the VCF special characters (e.g. TAB, :, ;, etc.) lying in the ASCII subset.  So you can treat the VCF as ASCII as parsing the but the embedded strings as UTF-8.  I think this is correct.

However, it specifies line termination as \r\n or \n.  That means the stop condition in the while loop is wrong.

Finally, the performance of this makes me sad, but it probably doesn't matter and we can fix it properly in C++ when we can use things like memchr.

Reply:
Performance makes me sad too. But until this is a bottleneck, I'm not worrying about it. I've pushed an update. @cseed please look again. I don't know what to do about dismissing a review when an approving review already exists.